storage/points_writer.go

```diff
@@ -8,13 +8,45 @@ import (
 	"github.com/influxdata/influxdb/v2"
 	"github.com/influxdata/influxdb/v2/kit/platform"
 	"github.com/influxdata/influxdb/v2/models"
+	"go.uber.org/zap"
 )
 
 // PointsWriter describes the ability to write points into a storage engine.
 type PointsWriter interface {
 	WritePoints(ctx context.Context, orgID platform.ID, bucketID platform.ID, points []models.Point) error
 }
 
+// MirroringPointsWriter sends all incoming points to both a Primary and Secondary writer.
+// Failures to write points to the Primary writer are bubbled up as errors. Failures to write
+// to the Secondary writer are logged, but not returned.
+type MirroringPointsWriter struct {
+	Primary   PointsWriter
+	Secondary PointsWriter
+	Log       *zap.Logger
+}
+
+func (w *MirroringPointsWriter) WritePoints(ctx context.Context, orgID platform.ID, bucketID platform.ID, p []models.Point) error {
+	if len(p) == 0 {
+		return nil
+	}
+
+	// Try writing to the primary first. If it fails, don't attempt to mirror to the secondary.
+	if err := w.Primary.WritePoints(ctx, orgID, bucketID, p); err != nil {
+		return err
+	}
+
+	// Mirror points to the secondary writer, logging on failure.
+	// Run the mirroring asynchronously to avoid adding too much overhead to the primary write path.
+	go func() {
```

Comment:
Agh, this probably won't be safe in all cases... for example the `to()` impl that will use this `PointsWriter` reuses a buffer of points. Will rework.